[dashboard] Framework for Heads-as-Processes.

rynewang opened the pull request:
This PR contains an architecture to support defining DashboardModules as subprocesses. It also contains an example TestModule in unit test.

When such a module is created, we have:

- a pair of `multiprocessing.Queue`: child_bound_queue, parent_bound_queue
- serializable message types for both directions
- child process: a module instance that is subclass of SubprocessModule
- child process: event loop
- child process: dedicated thread to listen from the child_bound_queue and dispatch messages to SubprocessModule
- parent process: a instance of SubprocessModuleHandle
- parent process: dedicated thread to listen from the parent_bound_queue and dispatch messages to SubprocessModuleHandle
- SubprocessRouteTable that provides decorators to define HTTP endpoints.

To accept subprocess modules:
- http_dashboard_head.py need to create SubprocessModuleHandle, one for each class type
- http_dashboard_head.py to bind SubprocessModuleHandle to SubprocessRouteTable
- http_dashboard_head.py to add SubprocessRouteTable to its aiohttp.web.Application

To define a module:
- create a new file (or modify existing one)
- define a MyModule as subclass of SubprocessModule
- define handlers as `async def my_method(self, request_body: bytes) -> aiohttp.web.Response:`
- use `@SubprocessRouteTable.get("/my_method")` to wrap handlers
- get registered by http_dashboard_head.py

On user end:
exactly the same

rynewang commented:
For an example, see https://github.com/ray-project/ray/pull/49732/files#diff-14a96417c0f4861db77ef9c11b73ed9c99dfd75b0611206180cfd31036016924R20
For how the message is transferred, see https://github.com/ray-project/ray/pull/49732/files#diff-10e3aeb5b09cfa47b40f75dd116e002f0081feb9c6f20c9cd5dc8aa222235b67R289

edoakes commented on python/ray/dashboard/subprocesses/message.py:
```diff
@@ -0,0 +1,79 @@
+from dataclasses import dataclass
+from typing import Union
+
+"""
+Child bound messages.
+"""
```
Did you consider just using gRPC instead of implementing a custom RPC framework?

edoakes commented on python/ray/dashboard/subprocesses/module.py:
```diff
@@ -0,0 +1,351 @@
+    config: dict
+
+
+class SubprocessModule(abc.ABC):
```
Did you consider using gRPC instead of implementing a custom RPC framework?

There are a lot of edge cases to get right here, the list of requirements looks very similar to an RPC framework to me, and we already use gRPC extensively throughout Ray.

jjyao replied:
I think we don't want to rely on python grpcio library as the quality of it is really bad and caused us lots of problems in the past. Our existing solution is using c++ grpc via cython but that doesn't apply here.

dentiny replied:
> we don't want to rely on python grpcio library as the quality of it is really bad and caused us lots of problems in the past

Completely not related to this PR, just curious what issues and problems did we meet in the past?

rynewang replied:
`grpcio` is a naughty python pkg. When Apple M1 first arrived, several versions of `grpcio` broke and we had to do something like `grpcio>base_version;grpcio!=1.2.3;grpcio!=1.2.4;grpcio!=1.2.5` and it's a headache

jjyao commented on python/ray/dashboard/subprocesses/module.py:
```diff
@@ -0,0 +1,351 @@
+        :param dashboard_head: The DashboardHead instance.
+        """
+        self._config = config
+        self._child_bound_queue = child_bound_queue
+        self._parent_bound_queue = parent_bound_queue
```
comments

jjyao replied:
Is my understanding correct: from the subprocess point of view, `child_bound_queue` is where the requests are and `parent_bound_queue` is where the responses are?

rynewang replied:
yes.

```
parent --(child_bound_queue of ChildBoundMessge)---> child
parent <-(parent_bound_queue of ParentBoundMessge)-- child
```

jjyao commented on python/ray/dashboard/subprocesses/module.py:
```diff
@@ -0,0 +1,351 @@
+    A Dashboard Head Module that runs in a subprocess. This is used with the decorators
+    to define a (request -> response) endpoint, or a (request -> AsyncIterator[bytes])
+    for a streaming endpoint.
+    """
```
can you document the thread model

rynewang replied:
sure. tldr: a thread running event loop, another thread listening to the child bound queue

rynewang commented on python/ray/dashboard/subprocesses/message.py:
```diff
@@ -0,0 +1,79 @@
+
+
+@dataclass
+class HealthCheckMessage:
```
I'm thinking to detect event loop hangs etc

rynewang commented on python/ray/dashboard/subprocesses/module.py:
```diff
@@ -0,0 +1,351 @@
+    Pickleable.
+    """
+
+    session_name: str
```
this is just a random field I came up about, and we should not have a typeless config anyway. I will remove `config` and add proper configs e.g. gcs_address when we onboard first real head.

jjyao commented on python/ray/dashboard/subprocesses/module.py:
```diff
@@ -0,0 +1,351 @@
+            asyncio.run_coroutine_threadsafe(
+                method(message, self._parent_bound_queue), loop
+            )
+        elif isinstance(message, HealthCheckMessage):
```
Can be implemented as normal RequestMessage

rynewang replied:
done

jjyao commented on python/ray/dashboard/subprocesses/module.py:
```diff
@@ -0,0 +1,351 @@
+
+    Before decoration:
+
+        @my_route.get("/get_logs")
```
```suggestion
        @SubprocessRouteTable.get("/get_logs")
```

jjyao commented on python/ray/dashboard/subprocesses/module.py:
```diff
@@ -0,0 +1,351 @@
+                handler = cls._decorated_non_streaming_handler(handler)
+
+            # Note we have 2 handlers:
+            # 1. the child side handler, that is `handler` that contains real logic and
```
how to handle timeout

rynewang replied:
I don't think we have timeout, at a aiohttp handler level, in the current dashboard. We can do it as aiohttp middleware so we don't need any code in `subprocesses/`. In this way, a handler may keep working, running to completion and got exceptions when returning a response, which should be ok for us.

rynewang commented on python/ray/dashboard/subprocesses/module.py:
```diff
@@ -0,0 +1,351 @@
+            asyncio.run_coroutine_threadsafe(
+                method(message, self._parent_bound_queue), loop
+            )
+        elif isinstance(message, HealthCheckMessage):
```
we don't have it right now; I plan to write it as a "fault tolerance PR" that includes:

1. periodic parent -> child health check req
2. if timed out, for several times (a policy is needed here), parent decides to consider the child dead
3. parent kills the proc; marks all active reqs dead
4. parent recreates a proc.

rynewang replied:
sure. actually Jiajun suggested we remove the HealthCheckMessage and just use RequestMessage to do the health check. will document it clearly

rynewang commented on python/ray/dashboard/subprocesses/runner.py:
```diff
@@ -0,0 +1,163 @@
+        self.loop = loop
+        self.module_cls = module_cls
+        self.config = config
+        self.child_bound_queue = multiprocessing.Queue()
+        self.parent_bound_queue = multiprocessing.Queue()
```
I don't want the queue to be blocking because it may get inserted in an async context. It should not be a problem anyway because the listener on both sides do not do heavy lifting: they just dispatch the msgs to the event loop. If later it do become a problem maybe we can do something like "when parent inserts msg to child bound queue, and the queue is full, directly fail the aiohttp request".

MengjinYan replied:
Make sense to leave it now and add mitigation when we see any issues as it is low risk. And at the same time, is there easy way for us to monitor whether it will become an issue?

rynewang replied:
yeah, I expect 99% of the time the queues are empty. we can add a metric on the queue size and e.g. alerts on size > 0 for a minute.

MengjinYan commented on python/ray/dashboard/subprocesses/module.py:
```diff
@@ -0,0 +1,351 @@
+        args=(loop,),
+        daemon=True,
+    )
+    dispatch_child_bound_messages_thread.start()
```
I'm mainly thinking of operations related to the queues. For example, having a clear signal of end of the queue to make sure all the messages in the queues are consumed.

rynewang replied:
hmm if the whole dashboard is shutdown do we really care all of messages are consumed and processed?

MengjinYan replied:
I see. I probably don't have the full understanding of how the processes are used. Yeah I think it's fine if the processes are only used by the dashboard and they are shutdown together with the dashboard.

alexeykudinkin replied:
@MengjinYan brings up a good point -- in case of graceful termination (say SIGTERM) we need to make sure we're not dropping anything on the floor (completing requests, finishing writing logs, etc) before wrapping up

rynewang commented:
Update:

1. Refactored. Now we have `handle.py`, `module.py`, `routes.py`, `message.py`, `utils.py` and `tests/`. It's much more clear.
2. Removed special HealthCheckMessage and its response message. Instead, reserve "_internal_health_check" API for health checks. One can call from the `SubprocessModuleHandle.health_check` to do it.
3. Added basic support for streaming responses. you can yield bytes in async methods now. No exception handling done yet (TODO).
4. fixed issues from pytest.
5. No timeout supports. I think we can postpone that until we need one.

MortalHappiness commented:
BTW, did you consider implement it with [ProcessPoolExecutor](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ProcessPoolExecutor)?

rynewang commented:
> BTW, did you consider implement it with [ProcessPoolExecutor](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ProcessPoolExecutor) and [run_in_executor](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor)? I think we don't need to maintain `Queue` and threads if we implement it in this way.

The problem is we need to keep states in the subprocess, and if we use ProcessPoolExecutor we will have to do that by defining globals and take care NOT to import them in the parent process which can be error prone and less intuitive.

rynewang commented on python/ray/dashboard/subprocesses/handle.py:
```diff
@@ -0,0 +1,240 @@
+):
+    """Handles a message from the parent bound queue."""
+    if isinstance(message, ResponseMessage):
+        response_fut = handle.active_requests[message.id].response_fut
```
makes sense, a better way is to always do work within the event loop. Will change

rynewang replied:
I found if you add all tasks to event loop you lose the order guarantees needed by the stream responses. Instead I made a thread safe dict.

rynewang commented on python/ray/dashboard/subprocesses/module.py:
```diff
@@ -0,0 +1,143 @@
+
+
+def run_module(
+    child_bound_queue: multiprocessing.Queue,
+    parent_bound_queue: multiprocessing.Queue,
```
wait no, multiprocessing.Queue does not support type annotations. reverted

rynewang replied:
```
(.conda) ~/gits/ray % python
Python 3.9.21 (main, Dec 11 2024, 10:21:40) 
[Clang 14.0.6 ] :: Anaconda, Inc. on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import multiprocessing
>>> q = multiprocessing.Queue()
>>> q2 = multiprocessing.Queue[int]()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'method' object is not subscriptable
>>> 
```

MortalHappiness replied:
Ah sorry, it's `queue.Queue` that supports generics.

rynewang commented on python/ray/dashboard/subprocesses/routes.py:
```diff
@@ -0,0 +1,220 @@
+                task = subprocess_module_handle.send_request(handler.__name__, request)
+                return await task
+
+            handler.__route_method__ = method
+            handler.__route_path__ = path
+            parent_side_handler.__route_method__ = method
+            parent_side_handler.__route_path__ = path
```
this is from prior art https://github.com/ray-project/ray/blob/5516b95fb0ce751d1d6537ac18c165deaa08bb77/python/ray/dashboard/optional_utils.py#L107 I will keep this until we need to do a big change for all Ray Dash.

edoakes commented:
> > BTW, did you consider implement it with [ProcessPoolExecutor](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ProcessPoolExecutor) and [run_in_executor](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor)? I think we don't need to maintain `Queue` and threads if we implement it in this way.
> 
> The problem is we need to keep states in the subprocess, and if we use ProcessPoolExecutor we will have to do that by defining globals and take care NOT to import them in the parent process which can be error prone and less intuitive.

A fairly simple solution to this would be to centralize all state in one class, then initialize that class using the `initializer` parameter to the [ProcessPoolExecutor](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ProcessPoolExecutor)

edoakes commented on python/ray/dashboard/subprocesses/handle.py:
```diff
@@ -0,0 +1,244 @@
+    def send_message(self, message: ChildBoundMessage):
+        self.child_bound_queue.put(message)
+
+    async def send_request(
+        self, method_name: str, request: Optional[aiohttp.web.Request]
+    ) -> Awaitable[aiohttp.web.Response]:
```
Does `aiohttp.web.Server` cancel the asyncio task when a client disconnects? If so we need to ensure we handle that gracefully here

edoakes commented on python/ray/dashboard/subprocesses/message.py:
```diff
@@ -0,0 +1,74 @@
+
+@dataclass
+class ResponseMessage:
+    id: str
```
is this an ID for the response or an ID for the request? please make the naming more clear

rynewang replied:
it's all request_id. updated naming

rynewang replied:
changed all to request_id

edoakes commented on python/ray/dashboard/subprocesses/handle.py:
```diff
@@ -0,0 +1,244 @@
+        try:
+            handle_parent_bound_message(loop, message, handle)
+        except Exception as e:
+            print(
```
Where does this `print` go? Need to ensure we have a solid logging story and this is observable

rynewang replied:
changed all to logging; added logging test. Now, a JobHead log goes to `dashboard-JobHead.log`. See https://github.com/ray-project/ray/pull/49732/files#diff-f70d196a48b75f61f76e7e9195fa60a69b2b00290f635b790ce714d14bb67ce8R143

edoakes replied:
awesome!

edoakes commented on python/ray/dashboard/subprocesses/handle.py:
```diff
@@ -0,0 +1,244 @@
+    """
+    assert_not_in_asyncio_loop()
+    while True:
+        message = parent_bound_queue.get()
```
suggest to also put this inside the `try` block as a failsafe

rynewang replied:
done

edoakes commented on python/ray/dashboard/subprocesses/tests/utils.py:
```diff
@@ -0,0 +1,53 @@
+        """
+        Streams the numbers 0 to N.
+        """
+        n = int(request_body)
+        for i in range(n):
+            await asyncio.sleep(0.001)
+            yield f"{i}\n".encode()
```
should add a test case that raises an exception midway through a streaming response

rynewang replied:
added

edoakes commented on python/ray/dashboard/subprocesses/tests/utils.py:
```diff
@@ -0,0 +1,53 @@
+    async def make_error_403(self, request_body: bytes) -> aiohttp.web.Response:
+        raise aiohttp.web.HTTPForbidden(reason="you shall not pass")
+
+    @SubprocessRouteTable.post("/streamed_iota", streaming=True)
+    async def streamed_iota(self, request_body: bytes) -> AsyncIterator[bytes]:
```
how do we set response headers for streaming handlers?

rynewang replied:
In this code you can not; it's hard coded in https://github.com/ray-project/ray/pull/49732/files#diff-9a1ab5ab72a357b774e76c3d7a74ffec29e120a8ed24c16eab82d076f9ffaefaR140 after MVP I think we can first yield a header dict then yield data chunks, and the framework use the header dict to set headers. do you think we need it right now?

edoakes commented on python/ray/dashboard/subprocesses/tests/utils.py:
```diff
@@ -0,0 +1,53 @@
+
+    @SubprocessRouteTable.put("/error_403")
+    async def make_error_403(self, request_body: bytes) -> aiohttp.web.Response:
+        raise aiohttp.web.HTTPForbidden(reason="you shall not pass")
```
:)

rynewang commented on python/ray/dashboard/subprocesses/handle.py:
```diff
@@ -0,0 +1,244 @@
+        )
+        self.dispatch_parent_bound_messages_thread.start()
+
+    def send_message(self, message: ChildBoundMessage):
+        self.child_bound_queue.put(message)
```
this is internal api. I think later we can provide other ChildBoundMessage types e.g. control messages, but for each one we should expose a dedicated Handle method that calls send_message. I will rename this to `_send_message`.

rynewang commented on python/ray/dashboard/subprocesses/handle.py:
```diff
@@ -0,0 +1,244 @@
+    handle: SubprocessModuleHandle,
+):
+    """Handles a message from the parent bound queue."""
+    if isinstance(message, ResponseMessage):
```
good point, adding

rynewang commented on python/ray/dashboard/subprocesses/handle.py:
```diff
@@ -0,0 +1,244 @@
+        request: aiohttp.web.Request
+        response_fut: Awaitable[aiohttp.web.Response]
+        # Only exists when the module decides this is a streaming response.
+        # To keep the data sent in order, we use future to synchronize. This assumes
+        # the Messages received from the Queue are in order.
+        # StreamResponseStartMessage expects this to be None. It creates the future,
+        # and in async, prepares a StreamResponse and resolves the future.
+        # StreamResponseDataMessage expects a future. It *replaces* the future with a
+        # new future by a coroutine that awaits the previous future, writes the data and
+        # resolves the new future.
+        # StreamResponseEndMessage expects a future. It resolves the future and sets
+        # the stream_response to None.
```
`asyncio.Queue` is also thread unsafe, so to do that you need to use `run_coroutine_threadsafe` to asynchronously put data into a `asyncio.Queue`, in which process you lose the order.

One way is to use a queue.Queue which is thread safe. Then in `__aiter__` use a thread pool executor to do blocking get. But this costs another thread and I don't know if this worths it.

edoakes replied:
I see, didn't think about `run_coroutine_threadsafe` losing ordering -- makes sense why you structured it this way now

rynewang replied:
If sleep(0) -> may risk busy waiting like a spinlock
if sleep(>0) -> may risk unnecessary delay introduced by the framework

can you elaborate on how we can better handle this dilemma?

alexeykudinkin replied:
I think 1ms sleep will hurt no one given that we have at most 1 hop b/w heads.

But now that i'm thinking about it a bit more, i think that thread-pool solution would probably be a more elegant one (just 1 thread, schedule the task that waits, once completes -- sets the future)

alexeykudinkin replied:
Will leave it up to you to decide if it's worth doing now or following up later given that you already have a working solution

rynewang commented on python/ray/dashboard/subprocesses/handle.py:
```diff
@@ -0,0 +1,244 @@
+        )
+        return await new_active_request.response_fut
+
+    async def health_check(self):
+        """
+        Do internal health check. The module should respond immediately with a 200 OK.
+        This can be used to measure module responsiveness in RTT, it also indicates
+        subprocess event loop lag.
+
+        Currently you get a 200 OK with body = b'ok!'. Later if we want we can add more
+        observability payloads.
+        """
+        return await self.send_request("_internal_health_check", request=None)
```
we can add periodical health checks on a layer above SubprocessModuleHandle. The dashboard parent process should manager each (cls -> SubprocessModuleHandle) and do restarts if needed.

rynewang commented on python/ray/dashboard/subprocesses/handle.py:
```diff
@@ -0,0 +1,244 @@
+        )
+        self.process.start()
+
+    def start_dispatch_parent_bound_messages_thread(self):
```
done

rynewang commented:
Update:

1. naming for fields and methods
2. used thread safe dict for the ActiveRequest look up.
3. added logging. Now, the module Config requires a bunch of logging related fields. These fields are meant to be populated with dashboard.py CLI arguments. It will mostly inherit whatever Dashboard config has: logging format, level, rotations. Only for the log filename it will amend with its own class name: `dashboard.log -> dashboard-JobHead.log`.
4. Metrics: no good stories here. we certainly don't want to expose N http ports, one for each Head. Currently we can accept not supporting metrics in subprocesses; because we don't have any in existing code anyway, other than MetricsHead. Later I imagine we can do an internal (rpc-only) API to expose any subprocess-level metrics, and when user call dashboard metrics API, we proxy that into a concatenation of all Head's metrics. But I think that will be something later than the MVP.
5. Streaming API support: I did some try but it's not trivial to write something like a ThreadSafeAsyncIterator that you can put from a listening thread, and consume from the event loop, either by `asyncio.Queue` or a `queue.Queue`. I suggest we just keep the current code, unless you have a better solution...

jjyao commented:
> Metrics: no good stories here. we certainly don't want to expose N http ports, one for each Head. Currently we can accept not supporting metrics in subprocesses; because we don't have any in existing code anyway, other than MetricsHead. Later I imagine we can do an internal (rpc-only) API to expose any subprocess-level metrics, and when user call dashboard metrics API, we proxy that into a concatenation of all Head's metrics. But I think that will be something later than the MVP.


We are doing that already in other places: core worker metrics are pushed to dashboard agent process first and let it expose to the outside world.

rynewang commented:
Added exception handling for streaming handlers. It will send 200 OK with the first data, then str(exception) then close the stream. Also added tests for it.

edoakes commented:
> Added exception handling for streaming handlers. It will send 200 OK with the first data, then str(exception) then close the stream. Also added tests for it.

Is the `str(exception)` sent all the way back to the client? Probably shouldn't be -- this is basically an internal server error

rynewang commented:
> > Added exception handling for streaming handlers. It will send 200 OK with the first data, then str(exception) then close the stream. Also added tests for it.
> 
> Is the `str(exception)` sent all the way back to the client? Probably shouldn't be -- this is basically an internal server error

do we have a reason not to send back detailed error message? There's no sensitive data and the code is open source anyway so maybe we can just send it for the sake of debugging.

rynewang commented:
TODO: declare SubprocessModule can only support non-minimal modules, for it requires aiohttp.

alexeykudinkin commented on python/ray/dashboard/subprocesses/module.py:
```diff
@@ -0,0 +1,177 @@
+    loop = asyncio.new_event_loop()
+    module = cls(config, child_bound_queue, parent_bound_queue)
+
+    loop.run_until_complete(module.run())
```
Wait, this is gonna block, right?

alexeykudinkin replied:
And it invokes run_forever internally, so you don't really need to do that below

rynewang replied:
This blocks until run() is done, this is intended to give Modules a chance to do init before it receives any incoming tasks

alexeykudinkin replied:
I see. Then let's rename it to `init` as `run` conventionally entails that this is the modules' main-loop

rynewang replied:
ok, renamed

rynewang commented on python/ray/dashboard/subprocesses/handle.py:
```diff
@@ -0,0 +1,295 @@
+       messages.
+    3. SubprocessRouteTable.bind(handle)
+    4. app.add_routes(routes=SubprocessRouteTable.bound_routes())
+    5. run the app.
```
done

rynewang commented on python/ray/dashboard/routes.py:
```diff
@@ -0,0 +1,194 @@
+                    bound_items.append(r)
+            else:
+                bound_items.append(r)
+        routes = aiohttp.web.RouteTableDef()
+        routes._items = bound_items
```
It's a code move from https://github.com/ray-project/ray/blob/16f0bb8ff1b7d60e4fddfbe1a45e42f4380983ba/python/ray/dashboard/optional_utils.py#L73 which is a move from older places, so this is an year old decision. I agree some day later we need to revisit this.

alexeykudinkin replied:
Got it. 

Would recommend next time to split out the code moving part into standalone PR so that it's easier to understand what's just moving and what's changing during the review

rynewang commented on python/ray/dashboard/routes.py:
```diff
@@ -0,0 +1,194 @@
+                        req = args[-1]
+                        return await handler(bind_info.instance, req)
+                    except Exception:
+                        logger.exception("Handle %s %s failed.", method, path)
```
I would prefer to keep it a pure code move from https://github.com/ray-project/ray/blob/16f0bb8ff1b7d60e4fddfbe1a45e42f4380983ba/python/ray/dashboard/optional_utils.py#L101; we can update later

rynewang commented:
TODO: to support cancellation

1. aiohttp server cancels coro when conn is broken.
2. For us, this means in the last line of `SubprocessModuleHandle.send_request`, `await new_active_request.response_fut` will raise `asyncio.CancelledError`.
3. We try-catch it, pop from `self.active_requests`, and send a child bound msg of new type `RequestCancelMessage(request_id=request_id)`
4. On child side, now we need to keep a {request_id -> task} map, just like the Handle side
5. ... and do task.cancel()

rynewang commented on python/ray/dashboard/subprocesses/tests/utils.py:
```diff
@@ -0,0 +1,74 @@
+        self.run_finished = False
+
+    async def run(self):
+        logger.info("TestModule is running")
+        self.run_finished = True
```
done

rynewang commented on python/ray/dashboard/subprocesses/routes.py:
```diff
@@ -0,0 +1,235 @@
+    4. SubprocessModule.dispatch_child_bound_messages receives the
+            RequestMessage and calls the child side handler.
+    (real work here)
+    5. child side handler sends a ParentBoundMessage to parent.
```
done

jjyao commented on python/ray/dashboard/routes.py:
```diff
@@ -0,0 +1,194 @@
+    """
+
+    class _BindInfo:
+        def __init__(self, filename, lineno, instance):
```
+1 let's annotation them

jjyao commented on python/ray/dashboard/routes.py:
```diff
@@ -0,0 +1,194 @@
+    """
+
+    class _BindInfo:
+        def __init__(self, filename, lineno, instance):
```
docstring, what's a BindInfo?

rynewang replied:
I did not want to change too much from this copy-and-paste PR. Let's do this in another place.

BindInfo is a mapping in routes[method][name] where instance is a `DashboardHeadModule | SubprocessModuleHandle`. filename and lineno is where this mapping is defined, to show a friendly error for duplicated (method, name) definitions.

jjyao commented on python/ray/dashboard/routes.py:
```diff
@@ -0,0 +1,194 @@
+    @classmethod
+    @property
+    @abc.abstractmethod
+    def _bind_map(cls):
```
docstring for all these methods

rynewang replied:
ditto, let's do this in another PR.

jjyao commented on python/ray/dashboard/subprocesses/handle.py:
```diff
@@ -0,0 +1,388 @@
+
+"""
+This file contains code run in the parent process. It can start a subprocess and send
+messages to it. Requires non-minimal Ray.
```
minimal heads won't be moved to subprocess?

rynewang replied:
*for now. Later I plan to add support for minimal subprocess modules. The plan is to separate minimal messages and non minimal messages in 2 different files, and have the optional handler to call non optional handler. We only have 2 optional modules so it's not so urgent

jjyao commented on python/ray/dashboard/subprocesses/handle.py:
```diff
@@ -0,0 +1,388 @@
+        self.process.start()
+
+        if start_dispatch_parent_bound_messages_thread:
+            self.dispatch_parent_bound_messages_thread = threading.Thread(
```
set thread name

rynewang replied:
done

jjyao commented on python/ray/dashboard/subprocesses/handle.py:
```diff
@@ -0,0 +1,388 @@
+        if self.process.exitcode is not None:
+            raise RuntimeError(f"Process exited with code {self.process.exitcode}")
+
+    async def _do_health_check_every_second(self):
```
why not calling it `_do_periodic_health_check` and pass in the frequency so if we later change from 1s to 2s, you don't need to change the filename lol

rynewang replied:
done

jjyao commented on python/ray/dashboard/subprocesses/handle.py:
```diff
@@ -0,0 +1,388 @@
+        response_fut: Awaitable[aiohttp.web.Response],
+    ) -> None:
+        """
+        When the async iterator in the module raises an error, we need to propagate it
```
is this how it's working now?

rynewang replied:
yes. we write `str(exception)` then write EOF to close the stream.

jjyao commented on python/ray/dashboard/subprocesses/handle.py:
```diff
@@ -0,0 +1,388 @@
+    message: ParentBoundMessage,
+    handle: SubprocessModuleHandle,
+):
+    """Handles a message from the parent bound queue."""
```
in the docstring, could you mention which thread will run this method

rynewang replied:
done

jjyao commented on python/ray/dashboard/subprocesses/handle.py:
```diff
@@ -0,0 +1,388 @@
+        try:
+            response = await asyncio.wrap_future(prev_fut)
+            await response.write_eof()
+            response_fut.set_result(response)
```
response is of type `StreamResponse` which is PARENT class of Response so we cannot set_result of StreamResponse if it's expecting Response.

rynewang replied:
done

jjyao reviewed (commented):
Will continue

jjyao commented on python/ray/dashboard/subprocesses/handle.py:
```diff
@@ -0,0 +1,390 @@
+    from the asyncio loop of the parent process.
+    """
+    assert_not_in_asyncio_loop()
+    while True:
```
We need a way to exit this thread otherwise if we start and destroy the same module multiple times, we will have thread leak?

rynewang replied:
done, added check

jjyao commented on python/ray/dashboard/subprocesses/routes.py:
```diff
@@ -0,0 +1,235 @@
+                    status=response.status,
+                    body=response.body,
+                )
+            except aiohttp.web.HTTPException as e:
+                # aiohttp.web.HTTPException cannot be pickled. Instead we send a
+                # UnaryResponseMessage with status and body.
+                reply_message = UnaryResponseMessage(
+                    request_id=message.request_id,
+                    status=e.status,
+                    body=e.text,
+                )
```
Is this handled on the parent side?

jjyao commented on python/ray/dashboard/subprocesses/routes.py:
```diff
@@ -0,0 +1,235 @@
+                task = subprocess_module_handle.send_request(handler.__name__, request)
+                return await task
+
+            handler.__route_method__ = method
+            handler.__route_path__ = path
+            parent_side_handler.__route_method__ = method
+            parent_side_handler.__route_path__ = path
```
Need some education here: isn't handler a child thing and parent_side_handler a parent thing, why are they initialized at the same time?

rynewang replied:
discussed offline.

jjyao commented on python/ray/dashboard/subprocesses/routes.py:
```diff
@@ -0,0 +1,235 @@
+    Before decoration:
+
+        @SubprocessRouteTable.get("/get_logs")
+        async def get_logs_method(self, request_body: bytes) \
```
should we convert from bytes to `aiohttp.web.Request`?

rynewang replied:
it's not easy. A aiohttp.web.Request is derived from a BaseRequest, which has fields like `payload_writer: AbstractStreamWriter` and is not intended for us to construct. Later when needed I will add `headers` and `params` to this method

jjyao commented on python/ray/dashboard/subprocesses/routes.py:
```diff
@@ -0,0 +1,235 @@
+                # aiohttp.web.HTTPException cannot be pickled. Instead we send a
+                # UnaryResponseMessage with status and body.
+                parent_bound_queue.put(
+                    UnaryResponseMessage(
```
this is for non-streaming mode

rynewang replied:
fixed to make sure we can send back a non-200 if we had never sent any data; and a error message if we already sent something.

rynewang commented on python/ray/dashboard/subprocesses/handle.py:
```diff
@@ -0,0 +1,391 @@
+        self.module_cls = module_cls
+        self.config = config
+
+        # Runtime states, set by start_module(), reset by destroy_module().
```
done

rynewang commented on python/ray/dashboard/subprocesses/handle.py:
```diff
@@ -0,0 +1,391 @@
+                # Empty is normal.
+                pass
+            except Exception as e:
+                logger.exception(
```
done

rynewang commented on python/ray/dashboard/subprocesses/module.py:
```diff
@@ -0,0 +1,191 @@
+                self.handle_child_bound_message(loop, message)
+            except Exception as e:
+                logger.exception(
+                    f"Error handling child bound message {message}. This request will hang forever."
```
we need to parse the message to get request_id and accordingly find the pending response, and we already do so in handle_parent_bound_message

rynewang commented on python/ray/dashboard/subprocesses/module.py:
```diff
@@ -0,0 +1,191 @@
+
+    async def _internal_health_check(
+        self, message: RequestMessage, parent_bound_queue: multiprocessing.Queue
+    ) -> None:
```
explained offline: we don't want a route assigned for it.

rynewang commented:
@jjyao if you approve I will merge this

jjyao commented on python/ray/dashboard/subprocesses/routes.py:
```diff
@@ -129,18 +130,22 @@ async def _streaming_handler(
-                        status=e.status,
-                        body=e.text,
+                if not start_message_sent:
+                    parent_bound_queue.put(
```
add test for it

rynewang replied:
added, it's test_streamed_error_before_yielding
